### PR TITLE
[RetroPlayer] allow using wayland with CRPRendererDMA

### PR DESCRIPTION
**Pull request by lrusak**
This allows us to use the `CRPRendererDMA` with wayland windowing (as long as there is registered `CBufferObjects`.

The new `CBufferObjects` that were merged in #17523 and #17532 allow wayland to use the DMA rendering path.

I added a commit to check if `CBufferObjectFactory` contains registered `CBufferObject` types so that we can fallback seemlessly to the `CRPRendererOpenGLES` if none are registered. @garbear please check this commit and let me know if you think that is the correct place to check this.

### CODE REVIEW

File: xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp [hunk 1]
```diff
@@ -11,17 +11,18 @@
...
-using namespace KODI::WINDOWING::GBM;
 using namespace KODI;
 using namespace RETRO;
 
 CRenderBufferDMA::CRenderBufferDMA(CRenderContext& context, int fourcc)
   : m_context(context),
     m_fourcc(fourcc),
     m_egl(new CEGLImage(
-        static_cast<CWinSystemGbmEGLContext*>(CServiceBroker::GetWinSystem())->GetEGLDisplay())),
+        dynamic_cast<KODI::WINDOWING::LINUX::CWinSystemEGL*>(CServiceBroker::GetWinSystem())
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp]: Please leave `static_cast()`:

* dynamic_cast can return nullptr, which makes f*cking up here a runtime error instead of a programming error, and programming errors can be easier to catch
* dynamic_cast has an extra RTTI hit

File: xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.cpp [hunk 2]
```diff
@@ -34,6 +35,9 @@ CRPBaseRenderer* CRendererFactoryDMA::CreateRenderer(const CRenderSettings& sett
 
 RenderBufferPoolVector CRendererFactoryDMA::CreateBufferPools(CRenderContext& context)
 {
+  if (!CBufferObjectFactory::HasRegisteredBufferObjects())
+    return {};
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.cpp]: > @garbear please check this commit and let me know if you think that is the correct place to check this.

It looks like returning empty here is the expected behavior: https://github.com/xbmc/xbmc/blob/4e57e432f23299ee78fcfa105f2d96989a7065e3/xbmc/cores/RetroPlayer/process/RPProcessInfo.cpp#L46

**Comment by garbear** [comment]
Tentative approval, let's wait for morning coffee to kick in and I'll be back later :)

File: xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp (hunk 1 above)
Comment (lrusak) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp]:
```
/home/lukas/Documents/git/xbmc/xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp: In constructor ‘KODI::RETRO::CRenderBufferDMA::CRenderBufferDMA(KODI::RETRO::CRenderContext&, int)’:
/home/lukas/Documents/git/xbmc/xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp:24:91: error: invalid static_cast from type ‘CWinSystemBase*’ to type ‘KODI::WINDOWING::LINUX::CWinSystemEGL*’
   24 |         static_cast<KODI::WINDOWING::LINUX::CWinSystemEGL*>(CServiceBroker::GetWinSystem())
      |                                                                                           ^
```

If you want I can break this line up in the constructor so we can check for errors

File: xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h [hunk 3]
```diff
@@ -19,7 +19,7 @@ namespace RETRO
 class CRendererFactoryDMA : public IRendererFactory
 {
 public:
-  ~CRendererFactoryDMA() override = default;
+  virtual ~CRendererFactoryDMA() override = default;
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h]: `virtual` shouldn't be required, right?

Comment (lrusak) [reply xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h]: right, I'm just not sure why the build is failing on jenkins

```
08:58:38 [100%] Linking CXX executable kodi-wayland
08:58:46 /home/jenkins/workspace/LINUX-64-Wayland/xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h:19: error: undefined reference to 'vtable for KODI::RETRO::CRendererFactoryDMA'
08:58:46 /usr/bin/ld.gold: the vtable symbol may be undefined because the class is missing its key function
```

Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h]: Try sticking the `= default` destructor bodies in .cpp files. Start with CRendererFactoryDMA.

File: xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp]: why is static_cast invalid? is CWinSystemEGL not a subclass of CWinSystemBase? maybe it's unhappy with the downcast?

Comment (lrusak) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp]: >CWinSystemEGL not a subclass of CWinSystemBase

this

File: xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h (hunk 3 above)
Comment (lrusak) [reply xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMA.h]: I think I figured it out :man_facepalming: jenkinsbuilds wayland-gl and I had it set to only build for wayland-gles

File: xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp]: can you explain how the dynamic_cast works? I haven't had a chance to do a comprehensive read yet

Comment (lrusak) [reply xbmc/cores/RetroPlayer/buffers/RenderBufferDMA.cpp]: `CWinSystemGbmEGLContext` inherits from `CWinSystemEGL` and `CWinSystemBase`
see :
https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/gbm/WinSystemGbmEGLContext.h#L26
and
https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/linux/WinSystemEGL.h#L20

File: xbmc/utils/BufferObjectFactory.h [hunk 4]
```diff
@@ -44,5 +44,5 @@ class CBufferObjectFactory
   static void ClearBufferObjects();
 
 protected:
-  static std::vector<std::function<std::unique_ptr<CBufferObject>()>> m_bufferObjects;
+  static std::list<std::function<std::unique_ptr<CBufferObject>()>> m_bufferObjects;
```
Comment (yol) [inline xbmc/utils/BufferObjectFactory.h]: I'm curious - why this change?

File: xbmc/utils/BufferObjectFactory.cpp [hunk 5]
```diff
@@ -8,17 +8,22 @@
...
-std::vector<std::function<std::unique_ptr<CBufferObject>()>> CBufferObjectFactory::m_bufferObjects;
+std::list<std::function<std::unique_ptr<CBufferObject>()>> CBufferObjectFactory::m_bufferObjects;
 
 std::unique_ptr<CBufferObject> CBufferObjectFactory::CreateBufferObject()
 {
-  return m_bufferObjects.back()();
+  for (const auto bufferObject : m_bufferObjects)
+  {
+    return bufferObject();
+  }
```
Comment (yol) [inline xbmc/utils/BufferObjectFactory.cpp]: I can sort of see what you did there, but this reads really unintuitively. What's wrong with the old code and adding a `.empty()` check?

File: xbmc/utils/BufferObjectFactory.cpp [hunk 6]
```diff
@@ -8,17 +8,22 @@
...
 
-std::vector<std::function<std::unique_ptr<CBufferObject>()>> CBufferObjectFactory::m_bufferObjects;
+std::list<std::function<std::unique_ptr<CBufferObject>()>> CBufferObjectFactory::m_bufferObjects;
 
 std::unique_ptr<CBufferObject> CBufferObjectFactory::CreateBufferObject()
 {
-  return m_bufferObjects.back()();
+  for (const auto bufferObject : m_bufferObjects)
+  {
+    return bufferObject();
```
Comment (yol) [inline xbmc/utils/BufferObjectFactory.cpp]: Maybe that's a stupid question and not related to this PR, but why can you have multiple factories when only the last registered one ever gets used anyway?

Comment (garbear) [reply xbmc/utils/BufferObjectFactory.cpp]: This makes sense to me, you use a stack to give priority to the most recent factory registered. Then in your registration class you pessimistically test for the existence of "worse" buffers before "better" ones.

Comment (yol) [reply xbmc/utils/BufferObjectFactory.cpp]: Actually, the comment clears it up: "In the future this may include some criteria for selecting a specific CBufferObject derived type. Currently it returns the CBufferObject implementation that was registered last."

@lrusak With your patch though, it returns the one registered *first*, am I correct?

Comment (a1rwulf) [reply xbmc/utils/BufferObjectFactory.cpp]: @yol I don't think so, he's using `m_bufferObjects.emplace_front(createFunc);` now and returns the first object he finds which is still the last that was registered.

Comment (yol) [reply xbmc/utils/BufferObjectFactory.cpp]: Makes sense!

**Comment by lrusak** [comment]
@garbear is the change in 31df19c what you had in mind?

**Comment by garbear** [comment]
Two suggestions:
* Change `if (!winSystemEGL)` to `if (winSystemEGL == nullptr)` for a type-safe comparison
* Improve the exception message. "dynamic_cast failed to cast to CWinSystemEGL" is obvious from the code, so maybe mention that RetroPlayer DRM is only implemented on EGL platforms, and/or possibly what went wrong to end up failing the check

**Comment by lrusak** [comment]
@garbear updated

**Review by garbear — APPROVED** [review]
looks good!
